Make Pod binding a subresource (and be generic to other types)

smarterclayton opened the pull request:
This changes the v1beta3 API for binding to make Binding a generic
object, and exposes `POST /pods/{name}/binding` alongside `POST /bindings`.

It also corrects bugs and duplicate code inside of api_installer related
to subresources and handling names.

Fixes #5000

smarterclayton committed:
Add a subbindings resource as /pods/{name}/binding

Allows POST to create a binding as a child. Also refactors internal
and v1beta3 Binding to be more generic (so that other resources can
support Bindings).

smarterclayton committed:
Simplify api_installer and setup methods

smarterclayton commented:
@bgrant0607 review the URL and changes to v1beta3 and we can find another reviewer for the code

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -901,13 +901,14 @@ type NamespaceList struct {
-	PodID string `json:"podID"`
-	Host  string `json:"host"`
+	// To is the object to bind to.
+	To ObjectReference `json:"to" description:"to is the object to bind to"`
```
Maybe `Target` rather than `To`? I had to read this line twice to parse it.

Also, this is the internal object. Do we want descriptions on these fields? We don't have any so far, and we plan to eliminate the json tags eventually.

smarterclayton replied:
Cut and paste error.  Target reads better.


> On Mar 4, 2015, at 5:29 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> >  	ObjectMeta `json:"metadata,omitempty"`
> >  
> > -	PodID string `json:"podID"`
> > -	Host  string `json:"host"`
> > +	// To is the object to bind to.
> > +	To ObjectReference `json:"to" description:"to is the object to bind to"`
> Maybe Target rather than To? I had to read this line twice to parse it.
> 
> Also, this is the internal object. Do we want descriptions on these fields? We don't have any so far, and we plan to eliminate the json tags eventually.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/client/pods.go:
```diff
@@ -92,3 +94,7 @@ func (c *pods) Update(pod *api.Pod) (result *api.Pod, err error) {
 }
+
+func (c *pods) Bind(binding *api.Binding) error {
+	return c.r.Post().Namespace(c.ns).Resource("pods").Name(binding.Name).Suffix("binding").Body(binding).Do().Error()
```
Maybe we should create a Subresource() routine in the client, as opposed to just using Suffix(), to reinforce the new pattern.

smarterclayton replied:
Agree.

----- Original Message -----
> > @@ -92,3 +94,7 @@ func (c *pods) Update(pod *api.Pod) (result *api.Pod, err
> > error) {
> >  	err =
> >  	c.r.Put().Namespace(c.ns).Resource("pods").Name(pod.Name).Body(pod).Do().Into(result)
> >  	return
> >  }
> > +
> > +func (c *pods) Bind(binding *api.Binding) error {
> > +	return
> > c.r.Post().Namespace(c.ns).Resource("pods").Name(binding.Name).Suffix("binding").Body(binding).Do().Error()
> 
> Maybe we should create a Subresource() routine in the client, as opposed to
> just using Suffix(), to reinforce the new pattern.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/5054/files#r25898897

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -142,7 +143,14 @@ func (r *BindingREST) New() runtime.Object {
 	binding := obj.(*api.Binding)
-	err = r.assignPod(ctx, binding.PodID, binding.Host)
+	// TODO: move me to a binding strategy
+	if len(binding.Target.Kind) != 0 && (binding.Target.Kind != "Node" && binding.Target.Kind != "Minion") {
```
How hard would it be to eliminate the separate etcd object and just write/read the pod? To make the separate object work, we really need atomic deletion of the pod and binding (or a background reconciliation loop, which I'd like to avoid).

smarterclayton replied:
I think it's

* kubelet needs to validate constraints and report failed
* scheduler / rc need to handle failed correctly
* we remove the atomic check and replace it with a query
* we remove setting the bindings and replace it with writing status.Host

----- Original Message -----
> > @@ -142,7 +143,14 @@ func (r *BindingREST) New() runtime.Object {
> >  // Create ensures a pod is bound to a specific host.
> >  func (r *BindingREST) Create(ctx api.Context, obj runtime.Object) (out
> >  runtime.Object, err error) {
> >  	binding := obj.(*api.Binding)
> > -	err = r.assignPod(ctx, binding.PodID, binding.Host)
> > +	// TODO: move me to a binding strategy
> > +	if len(binding.Target.Kind) != 0 && (binding.Target.Kind != "Node" &&
> > binding.Target.Kind != "Minion") {
> 
> How hard would it be to eliminate the separate etcd object and just
> write/read the pod? To make the separate object work, we really need atomic
> deletion of the pod and binding (or a background reconciliation loop, which
> I'd like to avoid).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/5054/files#r25899370

bgrant0607 replied:
We need to do this for host ports, out-of-disk, and other scenarios/resources, also.

@erictune @dchen1107 @rjnagal @vishh Is someone working on implementing this protocol?

bgrant0607 commented on pkg/master/master.go:
```diff
@@ -406,9 +404,10 @@ func (m *Master) init(c *Config) {
-		"bindings":    bindingStorage,
+		"pods":         podStorage,
+		"pods/status":  podStatusStorage,
+		"pods/binding": bindingStorage,
```
API and path LGTM.

bgrant0607 commented:
This LGTM as a fine first step.

smarterclayton committed:
Support SubResource call on Request

Allows better semantic use of Request when dealing with sub resources,
and allows clients to ignore ordering.  Supports multiple segments because
sub-resources are less tightly structured than regular resources.

smarterclayton commented:
Updated to support SubResource - all comments addressed.

bgrant0607 commented on pkg/runtime/interfaces.go:
```diff
@@ -45,6 +45,11 @@ type ObjectTyper interface {
 }
 
+// ObjectCreater contains methods for instantiating an object by kind and version.
+type ObjectCreater interface {
```
nit:Creator

smarterclayton replied:
I mostly did this to be consistent with ObjectTyper.  Defer until we clean up runtime?

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -939,16 +939,14 @@ type NamespaceList struct {
-	TypeMeta   `json:",inline"`
+	TypeMeta `json:",inline"`
+	// ObjectMeta describes the object that is being bound.
 	ObjectMeta `json:"metadata,omitempty"`
```
There should be descriptions on all v1beta3 fields now. This probably requires a rebase.

bgrant0607 replied:
Actually #5150 hasn't been merged yet.

smarterclayton commented:
Merge??? :)

bgrant0607 commented:
Sorry, LGTM.
